File: projeto-01-multiprotocol-python/python/MQTT/mqtt_client.py
import paho.mqtt.client as mqtt
import time

class MQTTClient:
    def __init__(self, host, port, username=None, password=None):
        # Initialize MQTT broker parameters
        self.host = host
        self.port = port
        self.connected = False

        # Create MQTT client instance
        self.client = mqtt.Client()

        # Login with credentialsif username and password:
        if username and password:
            self.client.username_pw_set(username, password)

        # Callback for connection event
        self.client.on_connect = self._on_connect
        self.client.on_disconnect = self._on_disconnect

        # Call back from mqtt subscribe
        self.client.on_message = self._on_message
        self.on_message_callback = None

    def connect(self):
        try:
            # Connect to MQTT broker
            self.client.connect(self.host, self.port)
            # Start loop for reconection
            self.client.loop_start()
        except Exception as e:
            print("Error connecting to MQTT broker: {}".format(e))

        # Delay to allow connection to establish
        time.sleep(0.5)

        # Check if the client is connected to the broker
        if not self.connected:
            print("Failed to connect to MQTT broker")

    def disconnect(self):
        try:
            # Disconnect from MQTT broker
            self.client.loop_stop()
            self.client.disconnect()
            self.connected = False
            print("Disconnected from MQTT broker")
        except Exception as e:
            print("Error disconnecting from MQTT broker: {}".format(e))

    def publish(self, topic, message):
        try:
            # Check if the client is connected to the broker before publishing
            if self.connected:
                # Send MQTT message
                result = self.client.publish(topic, message)
                if result.rc != mqtt.MQTT_ERR_SUCCESS:
                    print("Failed to publish message: {}".format(result.rc))
            else:
                print("Failed to publish message: Not connected to MQTT broker")
        except Exception as e:
            print("Error publishing message: {}".format(e))

    def subscribe(self, topic):
        try:
            # Check if the client is connected to the broker before publishing
            if self.connected:
                # Send MQTT message
                self.client.subscribe(topic)
            else:
                print("Failed to subscribe message: Not connected to MQTT broker")
        except Exception as e:
            print("Error subscribe message: {}".format(e))


    def _on_connect(self, client, userdata, flags, rc):
        if rc == 0:
            self.connected = True
            print("Connected to MQTT broker")
        else:
            print("Failed to connect to MQTT broker, return code: {}".format(rc))

    def _on_disconnect(self, client, userdata, rc):
        self.connected = False
        print(f"Disconnected from MQTT broker, rc={rc}")

    def _on_message(self, client, userdata, message):
        if self.on_message_callback:
            topic = message.topic
            payload = message.payload.decode()
            self.on_message_callback(topic, payload)

File: project-02-multiprotocol-c/src/MQTT/mqtt_client.c
#include <stdio.h>
#include <string.h>
#include "mqtt_client.h"
#include <unistd.h>

// Initialize MQTT parameters
void mqtt_init(MQTTClientWrapper *wrapper, const char *host, int port, const char *username, const char *password)
{
    strcpy(wrapper->host, host);
    wrapper->port = port;
    strcpy(wrapper->username, username);
    strcpy(wrapper->password, password);
    wrapper->connected = 0;
    wrapper->on_message_callback = NULL;
}

// Method that calls the method in main.c
static int on_message_arrived(void *context, char *topicName, int topicLen, MQTTClient_message *message) 
{
    // Convert context to MQTTClientWrapper
    MQTTClientWrapper *wrapper = (MQTTClientWrapper *)context;

    // Organize string
    char payload[500];
    int len = message->payloadlen;
    if (len > 499) len = 499;
    memcpy(payload, message->payload, len); // Used to ensure that the final value will be a string
    payload[len] = '\0';

    if (wrapper->on_message_callback != NULL) 
    {
        wrapper->on_message_callback(topicName, payload);
    }

    // Release memory arrea
    MQTTClient_freeMessage(&message);
    MQTTClient_free(topicName);
    return 1;
}

static void mqtt_reconnect_loop(MQTTClientWrapper *wrapper)
{
    int rc;

    while (1) {
        printf("Attempting to reconnect...\n");

        MQTTClient_destroy(&wrapper->client); // Release memory area
        rc = mqtt_connect(wrapper);

        if (rc == MQTTCLIENT_SUCCESS) {
            printf("Reconnected successfully\n");
            return;
        }

        printf("Reconnect failed, retrying in 5 seconds...\n");
        sleep(5);
        // This will lock the code while trying to reconnect
        // This is not a big problem for this project because 
        // this device doesn't have a internal buffer from PLCs
    }
}

static void on_connection_lost(void *context, char *cause)
{
    MQTTClientWrapper *wrapper = (MQTTClientWrapper *)context;

    printf("Connection lost: %s\n", cause);
    wrapper->connected = 0;

    mqtt_reconnect_loop(wrapper);
}

int mqtt_connect(MQTTClientWrapper *wrapper)
{
    // Get the standart structure for MQTT parameters
    MQTTClient_connectOptions conn_opts = MQTTClient_connectOptions_initializer;
    int rc;

    char address[150];
    sprintf(address, "tcp://%s:%d", wrapper->host, wrapper->port);

    // Create MQTT client
    MQTTClient_create(&wrapper->client, address, "c_client", MQTTCLIENT_PERSISTENCE_NONE, NULL);

    // MQTT parameters
    conn_opts.keepAliveInterval = 20;
    conn_opts.cleansession = 1;
    conn_opts.username = wrapper->username;
    conn_opts.password = wrapper->password;

    // Set method that will be triggered when receive MQTT message
    MQTTClient_setCallbacks(wrapper->client, wrapper, on_connection_lost, on_message_arrived, NULL);

    // Connect MQTT broker
    rc = MQTTClient_connect(wrapper->client, &conn_opts);
    if (rc != MQTTCLIENT_SUCCESS) 
    {
        printf("Failed to connect, return code %d\n", rc);
        wrapper->connected = 0;
        return rc;
    }

    printf("Connected to MQTT broker\n");
    wrapper->connected = 1;
    return MQTTCLIENT_SUCCESS;
}

void mqtt_disconnect(MQTTClientWrapper *wrapper) 
{
    MQTTClient_disconnect(wrapper->client, 10000);
    MQTTClient_destroy(&wrapper->client); // Release memory area
    wrapper->connected = 0;
    printf("Disconnected from MQTT broker\n");
}

int mqtt_publish(MQTTClientWrapper *wrapper, const char *topic, const char *payload) 
{
    MQTTClient_message pubmsg = MQTTClient_message_initializer; // Start parameters for MQTT publish
    MQTTClient_deliveryToken token;

    pubmsg.payload = (void *)payload;
    pubmsg.payloadlen = strlen(payload);
    pubmsg.qos = 0;
    pubmsg.retained = 0;

    MQTTClient_publishMessage(wrapper->client, topic, &pubmsg, &token);
    MQTTClient_waitForCompletion(wrapper->client, token, 5000);

    return 0;
}

int mqtt_subscribe(MQTTClientWrapper *wrapper, const char *topic) 
{
    int rc = MQTTClient_subscribe(wrapper->client, topic, 0);
    return rc;
}

void mqtt_set_message_callback(MQTTClientWrapper *wrapper, MessageCallback callback) 
{
    wrapper->on_message_callback = callback;
}